toolkit/tools/internal/pkggraph/pkggraph.go

```diff
@@ -1092,6 +1106,91 @@ func (g *PkgGraph) AddGoalNode(goalName string, packages []*pkgjson.PackageVer,
 		}
 	}
 
+	// Expand the goal node if requested
+	if extraLayers > 0 {
+		g.addGoalNodeLayers(goalNode, extraLayers)
+	}
+
+	return
+}
+
+// addGoalNodeLayers will expand a goal node by one layer. For example, if the goaled node is "foo" (ie the goal node
+// points to "foo") and extraLevels is 1, the goal node will now link to all nodes which depend on "foo" as well as
+// "foo" itself. A node is considered to depend on "foo" if it is a run node that has edges connecting it to "foo"
+// without any other run nodes in between.
+//
+//	i.e., if "foo_run" -> "foo_build" -> "<Some Meta Node>" -> "bar_run" -> "bar_build", and we are expanding from "bar"
+//	with layers=1, only "foo_run" will be added to the goal nodes since "foo_build" and "<Some Meta Node>" are not run nodes.
+func (g *PkgGraph) addGoalNodeLayers(goalNode *PkgNode, layers int) {
+	logger.Log.Debugf("Expanding goal node %s by %d layers", goalNode.GoalName, layers)
+
+	var expandedGoalNodes []*PkgNode
```

Comment:
I think we can go ahead and drop slices entirely in favour of "sets". Maps are passed by reference, so `getNextGoalLayer` can work directly on its input.

Reply:
Gave it a shot, its not straight forward from what I can see. Its hard to keep track of the state (ie what is the difference between a node we haven't seen one we have visited and are done with, or one we need to check still). I'm sure there is a way to do it but getting the recursive call to jump over build nodes right is not trivial from what I can see. It tends to either stop too early, or recurse indefinately, without additional meta data.

Reply:
Yeah, I see the complexity now, makes sense.